standalone version of backend stage1 truncation

portalesHEP opened the pull request:
#### PR description:

New wrapper (and updated processor) for standalone S1 truncation. 

#### PR validation:

tested with adapted version of `L1THGCalUtilities/test/testHGCalL1T_RelValV11_cfg.py`

portalesHEP committed:
standalone version of backend stage1 truncation

jbsauvan commented:
Thanks @portalesHEP 
A few general comments:
- It would probably be better to use and update the Processor class created in the previous PR, rather than creating a new one 
- The FPGA ids are more configuration data rather than input data, so could be passed through the configuration object 
- Standard exception should be avoided in CMSSW, so could rather return error codes, that can be interpreted by the wrapper which can then throw a CMSSW exception

portalesHEP committed:
Replacing processor from previous PR with new one + adapted configs accordingly

portalesHEP committed:
moved sector definition and fpga ID to configuration

portalesHEP committed:
Replaced std exceptions + removed forgotten cmssw dependance (+ some code cleaning)

portalesHEP committed:
fixed python config

portalesHEP commented:
> Thanks @portalesHEP A few general comments:
> 
>     * It would probably be better to use and update the Processor class created in the previous PR, rather than creating a new one
> 
>     * The FPGA ids are more configuration data rather than input data, so could be passed through the configuration object
> 
>     * Standard exception should be avoided in CMSSW, so could rather return error codes, that can be interpreted by the wrapper which can then throw a CMSSW exception

Thanks @jbsauvan ! I implemented the comments, and removed the last (I believe) CMSSW dependence from `HGCalStage1TruncationImpl_SA.h`.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalStage1TruncationConfig_SA.h:
```diff
@@ -0,0 +1,72 @@
+                           const double roz_max,
+                           const unsigned roz_bins,
+                           const std::vector<unsigned> max_tcs_per_bins,
+                           const std::vector<double> phi_edges)
```
Better to use (const) references to vectors to avoid copy

portalesHEP replied:
changed

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalStage1TruncationConfig_SA.h:
```diff
@@ -0,0 +1,72 @@
+                       const double roz_max,
+                       const unsigned roz_bins,
+                       const std::vector<unsigned> max_tcs_per_bins,
+                       const std::vector<double> phi_edges) {
```
Better to use (const) references to vectors to avoid copy

portalesHEP replied:
changed

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalStage1TruncationImpl_SA.h:
```diff
@@ -0,0 +1,41 @@
+
+  uint32_t packBin(unsigned roverzbin, unsigned phibin) const;
+  void unpackBin(unsigned packedbin, unsigned& roverzbin, unsigned& phibin) const;
+  unsigned phiBin(unsigned roverzbin, double phi, std::vector<double> phiedges) const;
```
Pass vector by const reference to avoid copy

portalesHEP replied:
changed

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalStage1TruncationWrapper.cc:
```diff
@@ -0,0 +1,110 @@
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "DataFormats/L1Trigger/interface/L1Candidate.h"
+#include "DataFormats/L1Trigger/interface/BXVector.h"
+#include "DataFormats/DetId/interface/DetId.h"
```
Are all these includes needed?

portalesHEP replied:
Indeed, they aren't, now keeping only what is actually used.

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalStage1TruncationWrapper.cc:
```diff
@@ -0,0 +1,110 @@
+    const l1thgcfirmware::HGCalTriggerCellSACollection& fpga_tcs_out,
+    const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs_original,
+    std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs_trunc) const {
+  for (unsigned itc = 0; itc < fpga_tcs_out.size(); ++itc) {
```
Could use a range-based loop

portalesHEP replied:
changed

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalStage1TruncationWrapper.cc:
```diff
@@ -0,0 +1,110 @@
+  convertCMSSWInputs(fpga_tcs, fpga_tcs_SA);
+
+  l1thgcfirmware::HGCalTriggerCellSACollection tcs_out_SA;
+  unsigned ec = theAlgo_.run(fpga_tcs_SA, theConfiguration_, tcs_out_SA);
```
Would be easier to understand what `ec` is if it is named `error` or similar.

portalesHEP replied:
Now called `error_code`

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalStage1TruncationWrapper.cc:
```diff
@@ -0,0 +1,110 @@
+}
+
+void HGCalStage1TruncationWrapper::configure(
+    const std::tuple<const edm::EventSetup&, const edm::ParameterSet&, const unsigned&, const uint32_t&>& configuration) {
```
Is the `ParameterSet` needed to be passed in the tuple?

portalesHEP replied:
Indeed, it's not, I removed it.

jbsauvan commented on L1Trigger/L1THGCal/python/hgcalBackEndLayer2Producer_cfi.py:
```diff
@@ -177,6 +177,12 @@
+    
+hgcalBackEndStage2SAtruncationProducer = cms.EDProducer(
+    "HGCalBackendLayer2Producer",
+    InputCluster = cms.InputTag('hgcalBackEndStage1ProducerSA:HGCalBackendStage1ProcessorSA'),
```
Shouldn't need the `SA` in the names?

portalesHEP replied:
This whole producer definition should have been removed in my previous commit. I removed it now.

jbsauvan commented on L1Trigger/L1THGCal/python/hgcalTriggerPrimitivesNew_cff.py:
```diff
@@ -20,10 +20,10 @@
 from Configuration.Eras.Modifier_phase2_hgcalV10_cff import phase2_hgcalV10
 from Configuration.Eras.Modifier_phase2_hgcalV11_cff import phase2_hgcalV11
-from L1Trigger.L1THGCal.customTriggerGeometry import custom_geometry_decentralized_V10, custom_geometry_decentralized_V11
+from L1Trigger.L1THGCal.customTriggerGeometry import custom_geometry_V10, custom_geometry_V11_Imp3
```
Thanks for fixing this, I missed them when updating the geometry customs

jbsauvan commented on L1Trigger/L1THGCal/src/backend/HGCalStage1TruncationImpl_SA.cc:
```diff
@@ -0,0 +1,104 @@
+HGCalStage1TruncationImplSA::HGCalStage1TruncationImplSA() {}
+
+unsigned HGCalStage1TruncationImplSA::run(const l1thgcfirmware::HGCalTriggerCellSACollection& tcs_in,
+                                          const l1thgcfirmware::Stage1TruncationConfig theConf,
```
Should pass the config by reference to avoid the copy

portalesHEP replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/backend/HGCalStage1TruncationImpl_SA.cc:
```diff
@@ -0,0 +1,104 @@
+unsigned HGCalStage1TruncationImplSA::phiBin(unsigned roverzbin, double phi, std::vector<double> phiedges) const {
+  unsigned phi_bin = 0;
+  if (roverzbin >= phiedges.size())
+    return 2;
```
Returning `2` as an error might be confusing and not future-proof (if for instance we want to add more bins in the future).  Returning a signed int, and using `-1` as error code could be better.

portalesHEP replied:
Good point, now returning `-1`

jbsauvan commented on L1Trigger/L1THGCalUtilities/python/customNtuples.py:
```diff
@@ -19,6 +19,16 @@ def custom_ntuples_stage1_truncation(process):
+        if ntuple.NtupleName=='HGCalTriggerNtupleHGCClusters' or \
+           ntuple.NtupleName=='HGCalTriggerNtupleHGCTriggerCells' or \
+           ntuple.NtupleName=='HGCalTriggerNtupleHGCMulticlusters':
+            ntuple.Clusters = cms.InputTag('hgcalBackEndStage1ProducerSA:HGCalBackendStage1ProcessorSA')
```
Do we need the `SA` in the names?

portalesHEP replied:
Same as for the stage 2 producer, this whole function is now removed.

portalesHEP committed:
passing vectors by reference, removing unused includes and custom functions + minor fixes
